hedera-node/hedera-app/src/main/java/com/hedera/node/app/workflows/handle/HandleWorkflow.java

```diff
@@ -611,30 +606,29 @@ private HandleOutput executeTopLevel(@NonNull final UserTxn userTxn, @NonNull fi
-                    // those changes already being visible in the FAB
-                    final var addressBookWritableStoreFactory = new WritableStoreFactory(
-                            userTxn.stack(), AddressBookService.NAME, userTxn.config(), storeMetricsService);
-                    addressBookHelper.adjustPostUpgradeNodeMetadata(
-                            networkInfo,
-                            userTxn.config(),
-                            addressBookWritableStoreFactory.getStore(WritableNodeStore.class));
-
-                    if (streamMode != RECORDS) {
-                        // Only externalize this if we are streaming blocks
-                        streamBuilder.exchangeRate(exchangeRateManager.exchangeRates());
-                        userTxn.stack().commitTransaction(streamBuilder);
-                    } else {
+                    // just took effect; it would be nice to unify the FAB and stake metadata in the future.
+                    // IMPORTANT:
+                    //   (1) These K/V changes must be committed directly to the state instead of
+                    //   being accumulated in the user stack in case it is rolled back.
+                    //   (2) The K/V changes must also be written immediately in their own block item,
+                    //   instead of being added to the base builder state changes, because if there is
+                    //   a preceding NodeStakeUpdate added later in executeTopLevel(), *its* state changes
+                    //   will come before the base builder's changes in the block stream
+                    final var writableTokenStates = state.getWritableStates(TokenService.NAME);
+                    final int maxPrecedingRecords = userTxn.config()
+                            .getConfigData(ConsensusConfig.class)
+                            .handleMaxPrecedingRecords();
+                    doStreamingKVChanges(
+                            writableTokenStates,
+                            // Ensure that even if the user txn has preceding children, these state changes still have
+                            // an earlier consensus time; since in fact they are, in fact, committed first
+                            userTxn.consensusNow().minusNanos(maxPrecedingRecords + 1),
+                            () -> stakeInfoHelper.adjustPostUpgradeStakes(
+                                    networkInfo,
+                                    userTxn.config(),
+                                    new WritableStakingInfoStore(writableTokenStates),
+                                    new WritableNetworkStakingRewardsStore(writableTokenStates)));
+                    if (streamMode == RECORDS) {
                         // Only update this if we are relying on RecordManager state for post-upgrade processing
                         blockRecordManager.markMigrationRecordsStreamed();
                         userTxn.stack().commitSystemStateChanges();
```

Comment:
`userTxn.stack().commitSystemStateChanges();` Should this be outside the if condition?